src/service/resources-impl.js

```diff
@@ -991,6 +1210,54 @@ export class ResourcesImpl {
    * @private
    */
   discoverWork_() {
+    if (this.intersectionObserver_) {
+      // With IntersectionObserver, we typically defer measurements to the
+      // intersection callback. However, we still need:
+      // 1. On viewport size changes (relayoutAll), apply sizes/media queries
+      //    AND remeasure elements. The latter makes sure that we call
+      //    onLayoutMeasure/onMeasureChanged e.g. for owner components to
+      //    reposition children.
+      // 2. Support on-demand measurements via Resource.requestMeasure.
+
+      // TODO(willchou): Do we need to build _all_ elements (instead of
+      // just near-viewport elements) on page-ready?
+
+      // Phase 1.
+      // We apply sizes/media query here before the first intersection callback
+      // so that the correct element size and hidden state will be measured
+      // by the observer (which avoids the need for a remeasure).
+      this.resources_.forEach(r => {
+        // NOT_LAID_OUT is the state after build() but before measure().
+        if (this.relayoutAll_ || r.getState() == ResourceState.NOT_LAID_OUT) {
+          // TODO(willchou): May need to add another ResourceState to avoid
+          // multiple invocations before the first intersection callback.
+          r.applySizesAndMediaQuery();
+          dev().fine(TAG_, 'apply sizes/media query:', r.debugid);
+        }
+      });
+
+      // Phase 2.
+      // Remeasures for viewport size changes (relayoutAll) and requestMeasure.
+      const toUnload = [];
+      this.resources_.forEach(r => {
```

Comment:
For-loops please. This will allocate a brand new closure on every pass.

Reply:
Done. Note that we'll have dramatically fewer passes now though. :)